Add ability to skip or transform page encoding statistics in Parquet metadata

etseidl opened the pull request:
# Which issue does this PR close?

- Closes #8518.

# Rationale for this change

This builds on #8763 to add options to either skip decoding of the page encoding statistics array, or transform it to a bitmask. This gets rid of the last of the heavy allocations in the metadata decoder.

# What changes are included in this PR?
Adds new options to `ParquetMetaDataOptions`. Also adds more metadata benchmarks.

# Are these changes tested?
Yes

# Are there any user-facing changes?
No, just adds new field to `ColumnChunkMetaData`.

etseidl commented:
excerpts from new benchmarks
```
decode parquet metadata time:   [15.050 µs 15.105 µs 15.164 µs]
decode metadata with schema
                        time:   [7.7038 µs 7.7286 µs 7.7561 µs]
decode metadata with stats mask
                        time:   [14.035 µs 14.100 µs 14.182 µs
decode metadata with skip PES
                        time:   [13.976 µs 14.016 µs 14.060 µs]
decode parquet metadata (wide)
                        time:   [54.013 ms 54.236 ms 54.468 ms]
decode metadata (wide) with schema
                        time:   [48.399 ms 48.562 ms 48.738 ms]
decode metadata (wide) with stats mask
                        time:   [44.912 ms 45.077 ms 45.253 ms]
decode metadata (wide) with skip PES
                        time:   [44.500 ms 44.616 ms 44.739 ms]
```
Skipping the stats is not any faster than turning them into a mask 😮.

etseidl commented on parquet/src/file/metadata/mod.rs:
````diff
@@ -1050,12 +1051,43 @@ impl ColumnChunkMetaData {
+    ///         })
+    /// }
+    /// ```
+    pub fn page_encoding_stats_mask(&self) -> Option<&EncodingMask> {
````
I wonder if this should be `data_page_encoding_stats_mask` (or just `data_page_encoding_stats`) to make it clear it only has the stats for data pages.

etseidl commented on parquet/src/file/metadata/options.rs:
```diff
@@ -48,11 +70,70 @@ impl ParquetMetaDataOptions {
     }
+
+    // with_encoding_stats_as_mask
+    add_mutator!(encoding_stats_as_mask, bool);
```
Now that I've done the macro, it saves all of 3 lines per instance. I'm fine with doing away with it.

etseidl replied:
I removed the macro and replaced the multiple setters for stats skipping with a policy enum

alamb commented on parquet/src/file/metadata/mod.rs:
```diff
@@ -1274,7 +1321,13 @@ impl ColumnChunkMetaDataBuilder {
     /// Sets page encoding stats for this column chunk.
     pub fn set_page_encoding_stats(mut self, value: Vec<PageEncodingStats>) -> Self {
-        self.0.encoding_stats = Some(value);
+        self.0.encoding_stats = Some(ParquetPageEncodingStats::Full(value));
```
It might be nice here in the comments to call out that setting the stats will override and mask that was set, and vice versa.

alamb commented on parquet/src/file/metadata/options.rs:
```diff
@@ -0,0 +1,191 @@
+    // The outer option acts as a global boolean, so if `skip_encoding_stats.is_some()`
+    // is `true` then we're at least skipping some stats. The inner `Option` is a keep
+    // list of column indicies to decode.
+    skip_encoding_stats: Option<Option<Arc<HashSet<usize>>>>,
```
Maybe a bitmask would be the better (but that can be a subsequent PR perhaps)

alamb commented on parquet/src/file/metadata/options.rs:
```diff
@@ -0,0 +1,191 @@
+#[derive(Default, Debug, Clone)]
+pub struct ParquetMetaDataOptions {
+    schema_descr: Option<SchemaDescPtr>,
+    encoding_stats_as_mask: bool,
```
I personally suggest we: file a ticket / PR to change the default in the next major release

alamb commented on parquet/src/file/metadata/options.rs:
```diff
@@ -29,6 +33,24 @@ use crate::schema::types::SchemaDescPtr;
 #[derive(Default, Debug, Clone)]
 pub struct ParquetMetaDataOptions {
     schema_descr: Option<SchemaDescPtr>,
+    encoding_stats_as_mask: bool,
```
It took me a little while to grok how encoding_stats_as_mask and skip_encoding_stats were related. But now I see they are mutually exclusive

I think maybe creating a enum rather than Option / Option /etc would be clearer to me but since it is an implementation detail this is also find

etseidl commented on parquet/src/file/metadata/options.rs:
```diff
@@ -48,11 +70,83 @@ impl ParquetMetaDataOptions {
+    ///
+    /// [`encoding_stats`]:
+    /// https://github.com/apache/parquet-format/blob/786142e26740487930ddc3ec5e39d780bd930907/src/main/thrift/parquet.thrift#L917
+    pub fn set_keep_encoding_stats(&mut self, keep: &[usize]) {
```
Another option to reduce bloat is to add `keep` as an optional argument to `set_skip_encoding_stats`. That might be less confusing.

etseidl commented:
Thanks for the review @alamb. I've now refactored a bit and introduced a policy enum for the encoding stats. I like this much more than the earlier API and feel it addresses my concerns about bloat. Please check it out when you have a moment. 🙏

etseidl commented on parquet/src/file/metadata/options.rs:
```diff
@@ -17,8 +17,53 @@
+
+impl ParquetStatisticsPolicy {
+    /// Create a `ParquetStatisticsPolicy` to skip all columns except those in `keep`.
+    pub fn skip_except(keep: &[usize]) -> Self {
```
Thanks for the help and patience 🙏. I've been wanting to get this right before forging ahead with other options.

etseidl commented:
Any objections to merging this now? @alamb?
